source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -278,6 +319,45 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
     return Http::FilterHeadersStatus::Continue;
   }
 
+  if (config_->responseCacheConfig().has_value()) {
+#ifdef CACHE_PERF_TEST
+    // Fill cache for testing
+    if (auth_header_str == "magic_fill_cache_for_testing") {
+      for (std::size_t i = 0; i < config_->responseCache<uint16_t>().getMaxCacheSize() - 10; ++i) {
+        std::string test_key = "test_key_" + std::to_string(i);
+        config_->responseCache<uint16_t>().Insert(test_key, 200);
+      }
+    }
+#endif
+    if (config_->responseCacheRememberBodyHeaders()) {
+      auto cached_response =
+          config_->responseCache<Filters::Common::ExtAuthz::Response>().Get(headers);
+      if (cached_response.has_value()) {
+        onCompleteSub(std::make_unique<Filters::Common::ExtAuthz::Response>(
+            std::move(cached_response.value())));
+        return Http::FilterHeadersStatus::StopIteration;
+      }
+    } else {
+      // Retrieve the HTTP status code from the cache
+      auto cached_status_code = config_->responseCache<uint16_t>().Get(headers);
+      if (cached_status_code.has_value()) {
+        if (*cached_status_code >= 200 && *cached_status_code < 300) {
+          // Any 2xx response is a success: let the request proceed
+          return Http::FilterHeadersStatus::Continue;
+        } else {
+          // Non-2xx response: reject the request
+          decoder_callbacks_->streamInfo().setResponseFlag(
+              StreamInfo::CoreResponseFlag::UnauthorizedExternalService);
+          decoder_callbacks_->sendLocalReply(
+              static_cast<Http::Code>(*cached_status_code), "Unauthorized", nullptr, absl::nullopt,
+              Filters::Common::ExtAuthz::ResponseCodeDetails::get().AuthzDenied);
+
+          return Http::FilterHeadersStatus::StopIteration;
+        }
```

Comment:
All of this logic belongs in the cache extension implementation.

The changes here (in the core ext_authz filter) should be the following:

Right before ext_authz gRPC call is made, pass the generated `CheckRequest` to the cache via `absl::optional<CheckResponse> cached = config_->responseCacheConfig().getCachedResponse(check_request);` If this has a value, `onComplete(cached.value()); return;`

When receiving a response from the ext_authz server, call `config_->responseCacheConfig().cacheResponse(check_response);`

All work to extract various headers, response codes, etc, should be done in your cache extension. The code here should be extremely minimal.